File: ios/demos/tdd-basics-demo/Sources/Calculator/Calculator.swift
import Foundation

/**
 * Calculator class demonstrating TDD principles
 *
 * This implementation follows the Red-Green-Refactor cycle:
 * 1. RED: Tests are written first and fail
 * 2. GREEN: Minimal code is written to pass tests
 * 3. REFACTOR: Code is improved without changing behavior
 */
public class Calculator {
    
    private var lastResult: Double = 0.0
    private var history: [String] = []
    
    public init() {}
    
    /**
     * Adds two numbers together
     * - Parameters:
     *   - a: First number
     *   - b: Second number
     * - Returns: Sum of a and b
     */
    public func add(_ a: Double, _ b: Double) -> Double {
        let result = a + b
        recordOperation("\(a) + \(b) = \(result)")
        lastResult = result
        return result
    }
    
    /**
     * Subtracts second number from first
     * - Parameters:
     *   - a: First number
     *   - b: Number to subtract
     * - Returns: Difference of a and b
     */
    public func subtract(_ a: Double, _ b: Double) -> Double {
        let result = a - b
        recordOperation("\(a) - \(b) = \(result)")
        lastResult = result
        return result
    }
    
    /**
     * Multiplies two numbers
     * - Parameters:
     *   - a: First number
     *   - b: Second number
     * - Returns: Product of a and b
     */
    public func multiply(_ a: Double, _ b: Double) -> Double {
        let result = a * b
        recordOperation("\(a) * \(b) = \(result)")
        lastResult = result
        return result
    }
    
    /**
     * Divides first number by second
     * - Parameters:
     *   - a: Dividend
     *   - b: Divisor
     * - Returns: Quotient of a divided by b
     * - Throws: CalculatorError.divisionByZero if divisor is zero
     */
    public func divide(_ a: Double, _ b: Double) throws -> Double {
        guard b != 0.0 else {
            throw CalculatorError.divisionByZero("ERROR: Division by zero is not allowed")
        }
        let result = a / b
        recordOperation("\(a) / \(b) = \(result)")
        lastResult = result
        return result
    }
    
    /**
     * Calculates percentage of a number
     * - Parameters:
     *   - value: The base value
     *   - percentage: The percentage to calculate
     * - Returns: Percentage of the value
     */
    public func calculatePercentage(of value: Double, percentage: Double) -> Double {
        let result = (value * percentage) / 100
        recordOperation("\(percentage)% of \(value) = \(result)")
        lastResult = result
        return result
    }
    
    /**
     * Gets the last calculated result
     * - Returns: Last operation result
     */
    public func getLastResult() -> Double {
        return lastResult
    }
    
    /**
     * Gets the calculation history
     * - Returns: Array of performed operations
     */
    public func getHistory() -> [String] {
        return history
    }
    
    /**
     * Clears the calculator memory and history
     */
    public func clear() {
        lastResult = 0.0
        history.removeAll()
    }
    
    /**
     * Records an operation in the history
     * - Parameter operation: String representation of the operation
     */
    private func recordOperation(_ operation: String) {
        history.append(operation)
        print("CALCULATION: \(operation)")
    }
    
    /**
     * Validates input for basic arithmetic operations
     * - Parameters:
     *   - a: First operand
     *   - b: Second operand
     * - Returns: true if inputs are valid
     */
    public func validateInputs(_ a: Double, _ b: Double) -> Bool {
        return !(a.isNaN || b.isNaN || a.isInfinite || b.isInfinite)
    }
}

/**
 * Errors that can be thrown by Calculator operations
 */
public enum CalculatorError: Error, LocalizedError {
    case divisionByZero(String)
    
    public var errorDescription: String? {
        switch self {
        case .divisionByZero(let message):
            return message
        }
    }
}

File: android/demos/tdd-basics-demo/src/main/kotlin/Calculator.kt
/**
 * Calculator class demonstrating TDD principles
 * 
 * This implementation follows the Red-Green-Refactor cycle:
 * 1. RED: Tests are written first and fail
 * 2. GREEN: Minimal code is written to pass tests
 * 3. REFACTOR: Code is improved without changing behavior
 */
class Calculator {
    
    private var lastResult: Double = 0.0
    private val history = mutableListOf<String>()
    
    /**
     * Adds two numbers together
     * @param a First number
     * @param b Second number  
     * @return Sum of a and b
     */
    fun add(a: Double, b: Double): Double {
        val result = a + b
        recordOperation("$a + $b = $result")
        lastResult = result
        return result
    }
    
    /**
     * Subtracts second number from first
     * @param a First number
     * @param b Number to subtract
     * @return Difference of a and b
     */
    fun subtract(a: Double, b: Double): Double {
        val result = a - b
        recordOperation("$a - $b = $result")
        lastResult = result
        return result
    }
    
    /**
     * Multiplies two numbers
     * @param a First number
     * @param b Second number
     * @return Product of a and b
     */
    fun multiply(a: Double, b: Double): Double {
        val result = a * b
        recordOperation("$a * $b = $result")
        lastResult = result
        return result
    }
    
    /**
     * Divides first number by second
     * @param a Dividend
     * @param b Divisor
     * @return Quotient of a divided by b
     * @throws IllegalArgumentException if divisor is zero
     */
    fun divide(a: Double, b: Double): Double {
        if (b == 0.0) {
            throw IllegalArgumentException("ERROR: Division by zero is not allowed")
        }
        val result = a / b
        recordOperation("$a / $b = $result")
        lastResult = result
        return result
    }
    
    /**
     * Calculates percentage of a number
     * @param value The base value
     * @param percentage The percentage to calculate
     * @return Percentage of the value
     */
    fun calculatePercentage(value: Double, percentage: Double): Double {
        val result = (value * percentage) / 100
        recordOperation("$percentage% of $value = $result")
        lastResult = result
        return result
    }
    
    /**
     * Gets the last calculated result
     * @return Last operation result
     */
    fun getLastResult(): Double = lastResult
    
    /**
     * Gets the calculation history
     * @return List of performed operations
     */
    fun getHistory(): List<String> = history.toList()
    
    /**
     * Clears the calculator memory and history
     */
    fun clear() {
        lastResult = 0.0
        history.clear()
    }
    
    /**
     * Records an operation in the history
     * @param operation String representation of the operation
     */
    private fun recordOperation(operation: String) {
        history.add(operation)
        println("CALCULATION: $operation")
    }
    
    /**
     * Validates input for basic arithmetic operations
     * @param a First operand
     * @param b Second operand
     * @return true if inputs are valid
     */
    fun validateInputs(a: Double, b: Double): Boolean {
        return !(a.isNaN() || b.isNaN() || a.isInfinite() || b.isInfinite())
    }
}
